core/src/main/java/io/grpc/Metadata.java

```diff
@@ -324,11 +339,96 @@ private void mergePathAndAuthority(Metadata other) {
     public String toString() {
       return "Headers(path=" + path
           + ",authority=" + authority
+          + ",timeout=" + timeout
           + ",metadata=" + super.toStringInternal() + ")";
     }
   }
 
   /**
+   * Container class for timeout information as represented in Headers.
+   */
+  public static class Timeout extends Metadata {
```

Comment:
Instead of creating a new class and adding getter/setter to `Headers`, create a `Metadata.Key` along with an `AsciiMarshaller`. For parseAsciiString, feel free to just `throw new UnsupportedOperationException`. For now, putting that `Metadata.Key`/`Marshaller` in `ChannelImpl` would probably make the most sense.

So `completeHeaders(...)` would become something more like:
```java
headers.removeAll(TIMEOUT_KEY);
if (headers.getTimeout() == null) {
  headers.put(TIMEOUT_KEY, method.getTimeout());
}
```

That would be similar to what is being done in the [transport elsewhere](https://github.com/grpc/grpc-java/blob/4ee2a6584a2f3a2054654409355b1f7995433b96/core/src/main/java/io/grpc/transport/AbstractServerStream.java#L120).

Note that I don't allow a user to set timeout themselves. That is probably the preferred behavior.

I don't expect changing the code based on this comment to be too hard; it seems like you could reuse most of the existing code.